Drivers: DMA: Infineon PSE84 Cat1 PDL driver

jsbatch opened the pull request:
Adding Infineon PDL based driver to support DMA on the PSOC Edge 84 microcontroller family.  Adds support for the following tests:
- tests/drivers/dma/chan_blen_transfer
- tests/drivers/dma/loop_transfer

jsbatch committed:
dts: infineon: pse84: Add bindings for ifx_cat1_dma_pdl driver

Adding binding file for IFX Cat1 DMA PDL based driver implementation.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
modules: hal_infineon: DMA cmake lists update

Updates DMA includes from the modules needed for DMA PDL based driver
implementation.
Cleans up cmake to use only USE_INFINEON_DMA Kconfig option to select
dma files from PDL/HAL.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
Drivers: DMA: Infineon PSE84: Adding DMA PDL based driver for pse84

Adds Infineon Cat1 PDL based driver for DMA.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
tests: drivers: dma: Add kit_pse84_eval test overlays

Adds overlays to enable test runs for DMA on the following tests:
 -tests/drivers/dma/chan_blen_transfer
 -tests/drivers/dma/loop_transfer

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch committed:
boards: Infineon: kit_pse84_eval: Adding DMA support

Adds support to the kit_pse84_eval board for DMA.

Signed-off-by: John Batch <john.batch@infineon.com>

JarmouniA commented on dts/bindings/dma/infineon,cat1-dma-pdl.yaml:
```diff
@@ -0,0 +1,216 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+title: Infineon CAT1 DMA PDL Driver
```
```suggestion
title: CAT1 DMA PDL
```

teburd commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,659 @@
+	data->channels[channel].error_callback_dis = config->error_callback_dis;
+
+	/* Lock and page in the channel configuration */
+	uint32_t key = irq_lock();
```
While this is ok, its not great to block interrupts while translating the descriptors. I get the reasoning here is in part because you have a single pool of descriptors of each dma that is shared among all channels. sys_mem_blocks could be used instead at a small cost to maintain the shared pool per dma instance.

jsbatch replied:
I've reworked things done within the irq_lock in an effort to minimize the amount of time.

teburd commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,659 @@
+			}
+
+			/* initialize descriptor */
+			dma_status = Cy_DMA_Descriptor_Init(descriptor, &descriptor_config);
```
It's unclear from the HAL usage here, but if registers are setup per channel then no locking at all should be needed. If some register is shared, a spin lock around the shared register updates is all that should be needed.

jsbatch replied:
As above, I've reworked the irq lock to avoid this.

teburd commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,659 @@
+	 * written back to the memory
+	 */
+#ifdef CONFIG_CPU_HAS_DCACHE
+	SCB_CleanDCache();
```
sys_cache_data_flush_all

jsbatch replied:
Updated to use sys_cache.

teburd reviewed (changes requested):
This looks like a duplicate of the existing dma_ifx_cat1.c driver to me. With maybe some added features. There shouldn't be two of these unless they are in fact two distinct IP blocks with different register maps and/or HAL interfaces.

The binding has a large number of options that don't appear to be used. These really should be added as they are required. By exposing these properties you are giving the appearance to someone creating a board port that these are configurable when they are not. That'll likely lead to some confusion.

teburd commented:
I understand now a bit better that this is an alternative driver to the cat1 dma IP built on a newer/redesigned HAL, that's fine. The issues still need to be addressed and the compatible should not change. A Kconfig selection should choose the driver variant and some documentation/comments about all of this should be done here to avoid confusion. I did not know what PDL means at first glance. Please add some comments in the driver file and Kconfig description noting this and add a zephyr issue to deprecate the older driver.

JarmouniA commented on dts/bindings/dma/infineon,cat1-dma-pdl.yaml:
See the discussion in https://github.com/zephyrproject-rtos/zephyr/pull/94264#discussion_r2413804751 for a way to switch to another driver, and deprecate the old.

jsbatch replied:
I've reworked this to use a Kconfig flag to differentiate between which devices use the older dma_ifx_cat1.c driver and the new dma_ifx_cat1_pdl.c driver.  The older driver uses an older HAL which is Infineon is phasing out.  The plan is for all devices to eventually migrate to the new PDL based driver, at which point the Kconfig flag will be removed.

jsbatch committed:
soc: infineon: Adding legacy hal definition to soc kconfig

Adds USE_INFINEON_LEGACY_HAL kconfig option to the PSOC6, CYW920829, and
xmc7200 series SOCs.  These devices have not transitioned to newer HAL
based drivers yet.

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,659 @@
+	 * cache and will lead to errors)
+	 */
+#ifdef CONFIG_CPU_HAS_DCACHE
+	SCB_InvalidateDCache();
```
I agree invalidating all of cache overkill, and we plan on addressing this in the future.  I've updated the function calls to use the sys_cache calls instead of calling into SCB directly.

JarmouniA commented on modules/hal_infineon/Kconfig:
```diff
@@ -9,6 +9,13 @@ config ZEPHYR_HAL_INFINEON_MODULE_BLOBS
+	help
+	  Use Legacy HAL driver instead of PDL based drivers.  This allows older devices to use HAL
+	  dased drivers until fully transitioned to the PDL.  The old HAL driver will be removed when 
+	  all devices have been transitioned.
```
```suggestion
	  Use Legacy HAL driver instead of PDL based drivers.  This allows devices to use HAL
	  based drivers until fully transitioned to the PDL.  The old HAL drivers will be removed when 
	  all devices have been transitioned.
```

jsbatch commented on soc/infineon/cat1a/Kconfig:
```diff
@@ -13,13 +13,15 @@ config SOC_FAMILY_PSOC6
 	select CPU_HAS_FPU
 	select SOC_FAMILY_INFINEON_CAT1
 	select BUILD_OUTPUT_HEX
+	select USE_INFINEON_LEGACY_HAL
```
Currently the PDL drivers are only used for the Edge (soc/infineon/edge) and PSOC C3 (soc/infineon/cat1b/psc3) devices.

jsbatch replied:
DMA hasn't yet been enabled on the PSOC C3.  Edge E84 is the device being added in this PR, and there are test overlays included. 

For legacy HAL, there are DMA ovelays to enable DMA testing on the cyw920829m2evk_02, cy8cproto_062_4242w, and xmc4x_relax_kit boards.

jsbatch commented:
> @jsbatch @billwatersiii out-of-scope for this PR, but the approach we followed here to introduce a PDL-based driver for the same hw device should be applied to all other infineon drivers, for ex. here
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/0f5e03f1fcba4326baf4507c343f3609bf32c524/drivers/serial/Kconfig.ifx_cat1#L23

Yes, I’m already working on updating other drivers to follow the same approach.  I wanted to get the approach finalized in this PR before getting too far into other driver changes.

jsbatch commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,565 @@
+	 * This detects concurrent configuration attempts on the same channel.
+	 * Also prevent reconfiguring while a transfer is active.
+	 */
+	uint32_t key = irq_lock();
```
I removed the status flags and locks.  I agree with you, after looking through the documentation I don't see anything indicating they are really necessary.

jsbatch commented on modules/hal_infineon/mtb-hal-cat1/CMakeLists.txt:
```diff
@@ -89,14 +89,17 @@ zephyr_library_sources_ifdef(CONFIG_USE_INFINEON_TIMER   ${hal_dir}/source/cyhal
-  zephyr_library_sources(${hal_dir}/source/cyhal_dma.c)
-  zephyr_library_sources(${hal_dir}/source/cyhal_dma_dmac.c)
-  zephyr_library_sources(${hal_dir}/source/cyhal_dma_dw.c)
+
```
I've updated the cmake lists to use the CONFIG_USE_INFINEON_DMA option for including these files.  That seems cleaner and makes it more clear.

teburd commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,523 @@
+	uint32_t i;
+	struct ifx_cat1_dma_data *data = dev->data;
+
+	for (i = 0u; i < CONFIG_INFINEON_DESCRIPTOR_POOL_SIZE; i++) {
```
There's a small race here though you aren't likely to hit it. Imagine getting interrupts just after you've done you branch on src == 0 and the interrupt steals this before you yourself get to change the marker.

Would recommend creating an atomic bit array and using this for your used/unused flag.

```c

/* When defining your driver intance and its data, create this atomic bit array
 * assign a pointer to the data struct... or have a global array for all dma instances, either works.
 */
ATOMIC_DEFINE(ifx_dma_desc_used_##inst, CONFIG_INFINEON_DESCRIPTOR_POOL_SIZE);

static cy_stc_dma_descriptor_t *ifx_cat1_dma_alloc_descriptor(const struct device *dev)
{
	uint32_t i;
	struct ifx_cat1_dma_data *data = dev->data;

	for (i = 0u; i < CONFIG_INFINEON_DESCRIPTOR_POOL_SIZE; i++) {
        if(!atomic_test_and_set_bit(data->desc_used, i)) {
            return &data->descriptor[i];
        }
    }
    
    return NULL;
}
```

jsbatch replied:
Good point, I've added that change.

jsbatch commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,535 @@
+void ifx_cat1_dma_free_descriptor(cy_stc_dma_descriptor_t *descr)
+{
+	if (descr != NULL) {
+		memset(descr, 0, sizeof(cy_stc_dma_descriptor_t));
```
That function isn't used anywhere, so I've removed it.

jsbatch commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -0,0 +1,528 @@
+	return 0;
+}
+
+cy_en_dma_data_size_t _convert_dma_data_size_z_to_pdl(struct dma_config *config)
```
Function updated.
